packages/wallets/saifu/src/adapter.ts

```diff
@@ -0,0 +1,270 @@
+    private _connecting: boolean;
+    private _wallet: SaifuWallet | null;
+    private _publicKey: PublicKey | null;
+    private _readyState: WalletReadyState =
+        typeof window === 'undefined' || typeof document === 'undefined'
+            ? WalletReadyState.Unsupported
+            : WalletReadyState.NotDetected;
+
+    constructor(config: SaifuWalletAdapterConfig = {}) {
+        super();
+        this._connecting = false;
+        this._wallet = null;
+        this._publicKey = null;
+
+        if (this._readyState !== WalletReadyState.Unsupported) {
+            scopePollingDetectionStrategy(() => {
+                if (window.solana?.isSaifu || window.saifu) {
+                    this._readyState = WalletReadyState.Installed;
+                    this.emit('readyStateChange', this._readyState);
+                    return true;
+                }
+                return false;
+            });
+        }
+    }
+
+    get publicKey(): PublicKey | null {
+        return this._publicKey;
+    }
+
+    get connecting(): boolean {
+        return this._connecting;
+    }
+
+    get connected(): boolean {
+        return !!this._wallet?.isConnected;
+    }
+
+    get readyState(): WalletReadyState {
+        return this._readyState;
+    }
+
+    async connect(): Promise<void> {
+        try {
+            if (this.connected || this.connecting) return;
+            if (this._readyState !== WalletReadyState.Installed) throw new WalletNotReadyError();
+
+            this._connecting = true;
+
+            const wallet = window.saifu || window.solana;
+
+            // just bail
+            if (!wallet) {
+                throw new WalletNotReadyError();
+            }
+
+            if (!wallet.isConnected) {
+                const handleDisconnect = wallet._handleDisconnect;
+                try {
+                    await new Promise<void>((resolve, reject) => {
+                        const connect = () => {
+                            wallet.off('connect', connect);
+                            resolve();
+                        };
+
+                        wallet._handleDisconnect = (...args: unknown[]) => {
+                            wallet.off('connect', connect);
+                            reject(new WalletWindowClosedError());
+                            return handleDisconnect.apply(wallet, args);
+                        };
+
+                        wallet.on('connect', connect);
+
+                        wallet.connect().catch((reason: any) => {
+                            wallet.off('connect', connect);
+                            reject(reason);
+                        });
+                    });
+                } catch (error: any) {
+                    if (error instanceof WalletError) throw error;
+                    throw new WalletConnectionError(error?.message, error);
+                } finally {
+                    wallet._handleDisconnect = handleDisconnect;
+                }
+            }
```

Comment:
The Saifu connection logic is inspired by how Phantom does things because when we built it we wanted maximum compatibility, so imitating the most popular wallet felt like the way to go

I cleaned up the `_handleDisconnect` hacks, but some of the code (like `wallet.on('connect')`) needs to stay until we can improve it on the Saifu side first

Reply:
Yeah, please don't imitate it, it's broken! The reason we access and override the private `_handleDisconnect` API is because it's broken. Changing the comments around this doesn't change that the whole thing is a hack.

This shouldn't be a part of the API at all. If you correctly reject the `connect` Promise when the extension window is closed by the user, you shouldn't have to handle it through events or private APIs.

Reply:
Ah, I thought with hack you meant how it keeps a reference to _handleDisconnect and then restores it back. I confirmed that we reject correctly on window close. Removed that part 👍